OrbitGl/ThreadTrack.cpp

```diff
@@ -138,31 +141,71 @@ void ThreadTrack::UpdateBoxHeight() {
 }
 
 bool ThreadTrack::IsEmpty() const {
-  return (GetNumTimers() == 0) && event_track_->IsEmpty() && tracepoint_track_->IsEmpty();
+  return thread_state_track_->IsEmpty() && event_track_->IsEmpty() &&
+         tracepoint_track_->IsEmpty() && (GetNumTimers() == 0);
+}
+
+void ThreadTrack::UpdatePositionOfSubtracks() {
+  const TimeGraphLayout& layout = time_graph_->GetLayout();
+  float thread_state_track_height = layout.GetThreadStateTrackHeight();
+  float event_track_height = layout.GetEventTrackHeight();
+  float space_between_subtracks = layout.GetSpaceBetweenTracksAndThread();
+
+  float current_y = pos_[1];
+
+  thread_state_track_->SetPos(pos_[0], current_y);
+  if (!thread_state_track_->IsEmpty()) {
+    current_y = current_y - space_between_subtracks - thread_state_track_height;
+  }
+
+  event_track_->SetPos(pos_[0], current_y);
+  if (!event_track_->IsEmpty()) {
+    current_y = current_y - space_between_subtracks - event_track_height;
+  }
+
+  tracepoint_track_->SetPos(pos_[0], current_y);
 }
 
 void ThreadTrack::Draw(GlCanvas* canvas, PickingMode picking_mode) {
   TimerTrack::Draw(canvas, picking_mode);
 
-  float event_track_height = time_graph_->GetLayout().GetEventTrackHeight();
+  UpdatePositionOfSubtracks();
 
-  event_track_->SetPos(pos_[0], pos_[1]);
-  event_track_->SetSize(canvas->GetWorldWidth(), event_track_height);
-  event_track_->Draw(canvas, picking_mode);
+  const TimeGraphLayout& layout = time_graph_->GetLayout();
+  float thread_state_track_height = layout.GetThreadStateTrackHeight();
+  float event_track_height = layout.GetEventTrackHeight();
+  float tracepoint_track_height = layout.GetEventTrackHeight();
 
-  tracepoint_track_->SetPos(pos_[0], pos_[1]);
-  tracepoint_track_->SetSize(canvas->GetWorldWidth(), event_track_height);
-  tracepoint_track_->Draw(canvas, picking_mode);
+  if (!thread_state_track_->IsEmpty()) {
+    thread_state_track_->SetSize(canvas->GetWorldWidth(), thread_state_track_height);
```

Comment:
Personally, I'd remove the gap between the thread state track and the timers, and make the thread state track that much taller, the thread states would be more visible and we wouldn't make the overall thread track height bigger, wdyt?